### PR TITLE
All applicants filter

### PR DESCRIPTION
**Pull request by nickdip**
Added functionality to show all applications.

### CODE REVIEW

File: app/lib/filters/show_all_applications.rb [hunk 1]
```diff
@@ -0,0 +1,23 @@
+module Filters
+  class ShowAllApplications < Base
+    def apply
+      if show_all_applications.empty?
```
Comment (thomasleese) [inline app/lib/filters/show_all_applications.rb:4]:
```suggestion
      unless show_all_applications.include?("show_all")
```

This value `show_all` comes from the value of the checkbox: https://github.com/DFE-Digital/apply-for-qualified-teacher-status/pull/2078/files#diff-14c048dcf48c873d6deb7ce9a22b554330cfadcac4a8e1c805b480f593b15983R69

I think your solution works too but I prefer this as it's more explicit and allows us to add more checkboxes in the future.

**Comment by nickdip** [comment]
I had to do some changes with the filtering, hopefuly it makes sense and works. I also ended up using a scope that removed draft applications

File: app/models/application_form.rb [hunk 2]
```diff
@@ -162,15 +162,7 @@ class ApplicationForm < ApplicationRecord
 
   STATUS_COLUMNS.each { |column| enum column, STATUS_VALUES, prefix: column }
 
-  scope :assessable, -> { where.not(stage: %i[draft completed]) }
```
Comment (thomasleese) [inline app/models/application_form.rb:165]: I think this scope is used in a number of other places so we'll need to keep it:

https://github.com/search?q=repo%3ADFE-Digital%2Fapply-for-qualified-teacher-status%20assessable&type=code

File: app/models/application_form.rb [hunk 3]
```diff
@@ -162,15 +162,7 @@ class ApplicationForm < ApplicationRecord
...
-
-  scope :active,
-        -> do
-          assessable
-            .or(where("awarded_at >= ?", 90.days.ago))
-            .or(where("declined_at >= ?", 90.days.ago))
-            .or(where("withdrawn_at >= ?", 90.days.ago))
-        end
+  scope :submitted, -> { where.not(stage: %i[draft]) }
```
Comment (thomasleese) [inline app/models/application_form.rb:165]:
```suggestion
  scope :submitted, -> { where.not(stage: "draft") }
```

Just a minor thing.

File: app/models/application_form.rb (hunk 2 above)
Comment (nickdip) [reply app/models/application_form.rb:165]: Oh yes...that's fixed now

File: app/views/assessor_interface/application_forms/index.html.erb [hunk 4]
```diff
@@ -63,6 +63,12 @@
         <% end %>
       <% end %>
     </section>
+
+    <section id="app-applications-show-all-applicants">
+      <%= f.govuk_check_boxes_fieldset :display, legend: { size: "s" }, small: true do %>
+        <%= f.govuk_check_box :display, :show_all, label: { text: "All applications" }, checked: @view_object.filter_form.display&.include?("show_all") %>
```
Comment (thomasleese) [inline app/views/assessor_interface/application_forms/index.html.erb:69]:
```suggestion
        <%= f.govuk_check_box :display, :show_all, label: { text: "Show applications completed older than 90 days ago" }, checked: @view_object.filter_form.display&.include?("show_all") %>
```

File: spec/view_objects/assessor_interface/application_forms_index_view_object_spec.rb [hunk 5]
```diff
@@ -149,8 +159,6 @@
   end
 
   describe "#country_filter_options" do
-    subject(:country_filter_options) { view_object.country_filter_options }
-
```
Comment (thomasleese) [inline spec/view_objects/assessor_interface/application_forms_index_view_object_spec.rb:153]: Did you mean to remove this line?

File: app/views/assessor_interface/application_forms/index.html.erb (hunk 4 above)
Comment (thomasleese) [inline app/views/assessor_interface/application_forms/index.html.erb:69]:
```suggestion
      <%= f.govuk_check_boxes_fieldset :display, legend: nil, small: true do %>
        <%= f.govuk_check_box :display, :show_all, label: { text: "All applications" }, checked: @view_object.filter_form.display&.include?("show_all") %>
```

I think we can hide the label here.

Comment (thomasleese) [inline app/views/assessor_interface/application_forms/index.html.erb:69]:
```suggestion
      <%= f.govuk_check_boxes_fieldset :show_all, legend: { size: "s" }, multiple: false, small: true do %>
        <%= f.govuk_check_box :show_all, :true, multiple: false, label: { text: "All applications" }, checked: @view_object.filter_form.show_all == "true" %>
```

Thinking about the naming of these fields, I think it makes sense to stick with the `show_all` name as `display` is a little too generic in my opinion, unclear exactly what that might be.

File: app/lib/filters/show_all_applications.rb [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+      end
+
+      scope
+    end
+
+    private
+
+    def show_all_applications
+      Array(params[:display]).compact_blank
+    end
```
Comment (thomasleese) [inline app/lib/filters/show_all_applications.rb:23]: Following on from my suggestion about renaming the check box name.

```suggestion
    def show_all?
      ActiveModel::Type::Boolean.new.cast(params[:show_all])
    end
```

`ActiveModel::Type::Boolean.new.cast` parses the string for us so the string `"true"` is treated as a `true` boolean.

File: app/lib/filters/show_all_applications.rb [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+      unless show_all_applications.include?("show_all")
+        ninety_days_ago = 90.days.ago
+        new_scope =
+          scope.where(
+            "(awarded_at >= :ninety_days OR awarded_at IS NULL)
+          AND (declined_at >= :ninety_days OR declined_at IS NULL)
+          AND (withdrawn_at >= :ninety_days OR withdrawn_at IS NULL)",
+            ninety_days: ninety_days_ago,
+          )
+        return new_scope
```
Comment (thomasleese) [inline app/lib/filters/show_all_applications.rb:13]:
```suggestion
        return scope.where(
            "(awarded_at >= :ninety_days OR awarded_at IS NULL)
          AND (declined_at >= :ninety_days OR declined_at IS NULL)
          AND (withdrawn_at >= :ninety_days OR withdrawn_at IS NULL)",
            ninety_days: ninety_days_ago,
          )
```

I don't think we need to define a variable here.

**Review by thomasleese — COMMENTED** [review]
Overall looks good, just get a couple of small comments left.

File: app/views/assessor_interface/application_forms/index.html.erb (hunk 4 above)
Comment (nickdip) [reply app/views/assessor_interface/application_forms/index.html.erb:69]: Makes sense - I think I'll need to adapt this in the spec too

File: spec/view_objects/assessor_interface/application_forms_index_view_object_spec.rb (hunk 5 above)
Comment (nickdip) [reply spec/view_objects/assessor_interface/application_forms_index_view_object_spec.rb:153]: no!

File: app/models/application_form.rb [hunk 8]
```diff
@@ -164,13 +164,9 @@ class ApplicationForm < ApplicationRecord
...
   scope :assessable, -> { where.not(stage: %i[draft completed]) }
 
-  scope :active,
-        -> do
-          assessable
-            .or(where("awarded_at >= ?", 90.days.ago))
-            .or(where("declined_at >= ?", 90.days.ago))
-            .or(where("withdrawn_at >= ?", 90.days.ago))
-        end
+  scope :assessable, -> { where.not(stage: %i[draft completed]) }
```
Comment (thomasleese) [inline app/models/application_form.rb:167]: It looks like we've got two `assessable` scopes defined now.

**Review by thomasleese — COMMENTED** [review]
One comment left otherwise it looks good!

File: config/feature_flags.yml [hunk 9]
```diff
@@ -34,4 +34,9 @@ feature_flags:
       Allow starting an application on this service directly after completing
       an eligibility check.
 
+  show_all_applicants:
```
Comment (thomasleese) [inline config/feature_flags.yml:37]:
```suggestion
  show_all_applications_filter:
```

Just a suggestion to make the name a little more explicit.

**Review by thomasleese — APPROVED** [review]

**Pull request merged**